### PR TITLE
BOOKKEEPER-588 SSL Support for Bookkeeper

### PR DESCRIPTION
**Pull request by kishorekasi**
+ Merged changes from @eoliville
+ Mutual Authentication

### CODE REVIEW

File: bookkeeper-server/conf/bk_server.conf [hunk 1]
```diff
@@ -317,3 +317,60 @@ zkTimeout=10000
 
 # Stats Provider Class
 #statsProviderClass=org.apache.bookkeeper.stats.CodahaleMetricsProvider
+
+<<<<<<< HEAD
```
Comment (eolivelli) [inline bookkeeper-server/conf/bk_server.conf]: This seems a problem with conflict resolution

File: bookkeeper-server/pom.xml [hunk 2]
```diff
@@ -222,8 +222,21 @@
         <version>2.7.3</version>
         <scope>test</scope>
     </dependency>          
+    <dependency>
+      <groupId>io.netty</groupId>
+      <artifactId>netty-tcnative-boringssl-static</artifactId>
+      <version>${netty-boringssl.version}</version>
+      <classifier>${os.detected.classifier}</classifier>
```
Comment (eolivelli) [inline bookkeeper-server/pom.xml]: We should use netty-tcnative-boringssl-static without classifier which contains native bindings for every os

File: bookkeeper-server/pom.xml [hunk 3]
```diff
@@ -222,8 +222,21 @@
...
+    </dependency>
   </dependencies>
   <build>
+    <extensions>
+      <extension>
+        <groupId>kr.motd.maven</groupId>
```
Comment (eolivelli) [inline bookkeeper-server/pom.xml]: Without the classifier for netty ssl this plugin will not be needed

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java [hunk 4]
```diff
@@ -1118,4 +1131,126 @@ public ClientConfiguration setClientRole(String role) {
...
+    public String getSSLTrustStoreType() {
+        return getString(SSL_TRUSTSTORE_TYPE, "JKS");
+    }
+
+    /**
+     * Set the truststore type for client.
+     * 
+     * @return
+     */
+    @VisibleForTesting
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java]: IMHO This annotation is to be removed

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java [hunk 5]
```diff
@@ -1118,4 +1131,126 @@ public ClientConfiguration setClientRole(String role) {
...
+    public String getSSLTrustStore() {
+        return getString(SSL_TRUSTSTORE, null);
+    }
+
+    /**
+     * Set the truststore path for the client.
+     * 
+     * @return
+     */
+    @VisibleForTesting
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java]: Same as above and for other new methods

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieRequestProcessor.java [hunk 6]
```diff
@@ -239,21 +279,53 @@ private void processReadRequestV3(final BookkeeperProtocol.Request r, final Chan
...
-        } else {
-            readThreadPool.submit(readLac);
+                response.setStatus(BookkeeperProtocol.StatusCode.EOK);
+                BookkeeperProtocol.StartTLSResponse.Builder builder = BookkeeperProtocol.StartTLSResponse.newBuilder();
+                response.setStartTLSResponse(builder.build());
+                sslHandler.handshakeFuture().addListener(new GenericFutureListener<Future<Channel>>() {
+                    @Override
+                    public void operationComplete(Future<Channel> future) throws Exception {
+                        // notify the AuthPlugin the completion of the handshake, even in case of failure
+                        LOG.info("Bookie STARTTLS handshake complete");
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieRequestProcessor.java]: Maybe this should be changed to debug level

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java [hunk 7]
```diff
@@ -1801,4 +1917,107 @@ public String toString() {
             return String.format("%d:%d %s", ledgerId, entryId, operationType);
         }
     }
+    
+    //public class ConnectionFutureListener implements ChannelFutureListener {
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java]: Remove comment

File: pom.xml [hunk 8]
```diff
@@ -156,6 +158,17 @@
   </reporting>
   <profiles>
     <profile>
+      <id>m2e</id>
```
Comment (eolivelli) [inline pom.xml]: This profile can be dropped

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookkeeperProtocol.java.old [hunk 9]
```diff
@@ -0,0 +1,14834 @@
+// Generated by the protocol buffer compiler.  DO NOT EDIT!
+// source: src/main/proto/BookkeeperProtocol.proto
```
Comment (fpj) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookkeeperProtocol.java.old]: This .old file should not be part of the PR, yes?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java [hunk 10]
```diff
@@ -331,6 +380,12 @@ protected ChannelFuture connect() {
             protected void initChannel(Channel ch) throws Exception {
                 ChannelPipeline pipeline = ch.pipeline();
 
+                // PerChannelBookieClient parentObj = PerChannelBookieClient.this;
```
Comment (fpj) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java]: Why is this code commented out? Should it be removed?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/ssl/SSLContextFactory.java [hunk 11]
```diff
@@ -0,0 +1,303 @@
+package org.apache.bookkeeper.ssl;
```
Comment (fpj) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/ssl/SSLContextFactory.java]: Package declaration should be after license header.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/ssl/SSLContextFactory.java [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.TrustManagerFactory;
+
+public class SSLContextFactory implements SecurityHandlerFactory {
```
Comment (fpj) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/ssl/SSLContextFactory.java]: This is a public class, shouldn't it have javadocs?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/ssl/SecurityException.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
+package org.apache.bookkeeper.ssl;
```
Comment (fpj) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/ssl/SecurityException.java]: Package declaration after license header, please.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/ssl/SecurityHandlerFactory.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
+package org.apache.bookkeeper.ssl;
```
Comment (fpj) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/ssl/SecurityHandlerFactory.java]: Package declaration after license header, please.

File: bookkeeper-server/src/main/proto/BookkeeperProtocol.proto [hunk 15]
```diff
@@ -60,6 +60,7 @@ enum OperationType {
     WRITE_LAC = 6;
     READ_LAC = 7;
     GET_BOOKIE_INFO = 8;
+    STARTTLS = 9;
```
Comment (fpj) [inline bookkeeper-server/src/main/proto/BookkeeperProtocol.proto]: To follow the convention of the previous operations, should we call it `START_TLS`?

**Comment by eolivelli** [comment]
I wonder if we have to add some "stat" about SSL clients. Maybe we can enhance this aspect in the future

File: bookkeeper-server/src/test/resources/create.certificates.sh [hunk 16]
```diff
@@ -0,0 +1,33 @@
+#/bin/bash
```
Comment (fpj) [inline bookkeeper-server/src/test/resources/create.certificates.sh]: What´s the purpose of this script? Testing? I'm wondering whether this is something we need to check in.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/ssl/SecurityProviderFactoryFactory.java [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+            throws SecurityException {
+        if ((securityHandler == null) || (securityHandler.equals(""))) {
+            return null;
+        }
+
+        ClassLoader classLoader = SecurityProviderFactoryFactory.class.getClassLoader();
+        SecurityHandlerFactory shFactory = null;
+        try {
+            Class<?> shFactoryClass = classLoader.loadClass(securityHandler);
+            shFactory = (SecurityHandlerFactory) shFactoryClass.getConstructor(AbstractConfiguration.class)
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/ssl/SecurityProviderFactoryFactory.java]: we should use ReflectionUtils.newInstance as in other parts of BookKeeper code.
At the same time if we want to keep actual style for dynamically loaded classes we should require a no-arg constructor and call an "initialize(conf)" method

File: bookkeeper-server/pom.xml (hunk 3 above)
Comment (fpj) [inline bookkeeper-server/pom.xml]: In the case we keep this plugin, do we need to worry about its license?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/ssl/SecurityHandlerFactory.java [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+import java.util.Date;
+
+import io.netty.channel.ChannelInboundHandlerAdapter;
+
+public interface SecurityHandlerFactory {
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/ssl/SecurityHandlerFactory.java]: IMHO As this class is tied to Netty4 we should warn in the docs (javadocs of the class) that it used only internally, and discourage "custom" implementations

File: bookkeeper-server/pom.xml (hunk 3 above)
Comment (eolivelli) [inline bookkeeper-server/pom.xml]: @fpj I think we will drop the plugin from this PR but we will need it for another issue, see #80 and my additional commit (that @sijie is going to append) https://github.com/eolivelli/bookkeeper/commit/4c7e571f48f5f7c713caa27d840b94f152f5b7ce 

it seems to me that it is licensed with Apache2 license

**Comment by eolivelli** [comment]
Overall is ok,
 but rat is failing and on jenkins one of the new tests failed, the mixed cluster one

Can you fix it?

File: pom.xml [hunk 19]
```diff
@@ -42,8 +42,9 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
     <guava.version>20.0</guava.version>
-    <netty.version>4.1.10.Final</netty.version>
     <protobuf.version>2.6.1</protobuf.version>
+    <netty.version>4.1.10.Final</netty.version>
```
Comment (eolivelli) [inline pom.xml]: @kishorekasi 
There are blocker issues reported on OpenSSL support on this version (4.1.10)
we should upgrade to 4.1.12 + 2.0.3final
http://netty.io/news/2017/05/12/4-0-47-Final-4-1-11-Final.html
https://netty.io/news/2017/06/09/4-0-48-Final-4-1-12-Final.html

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieRequestProcessor.java [hunk 20]
```diff
@@ -289,38 +293,33 @@ private void processStartTLSRequestV3(final BookkeeperProtocol.Request r, final
...
+            response.setStartTLSResponse(builder.build());
+            sslHandler.handshakeFuture().addListener(new GenericFutureListener<Future<Channel>>() {
+                @Override
+                public void operationComplete(Future<Channel> future) throws Exception {
+                    // notify the AuthPlugin the completion of the handshake, even in case of failure
+                    AuthHandler.ServerSideHandler authHandler = c.pipeline()
+                            .get(AuthHandler.ServerSideHandler.class);
+                    authHandler.authProvider.onProtocolUpgrade();
+                    if (future.isSuccess()) {
+                        LOG.info("Session is protected by: {}", sslHandler.engine().getSession().getCipherSuite());
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieRequestProcessor.java]: maybe this could be 'debug' and not 'info' (it was a debug for the auth plugin integration)

**Comment by kishorekasi** [comment]
Updated the pull request. I am not able to reproduced the previous jenkins failure locally. With this new update I will wait for the jenkins run to check if the failures are repeatable.

**Comment by eolivelli** [comment]
@sijie for me this patch is OK, I don't know if other developers/committer want to add other comments

@kishorekasi can you change the description of the PR which usually it written to "git log" to remove the (wrong) reference to my github account and add a reference to @ivankelly which was the original author of the patch ?

We will need to add the documentation to the security docs, maybe we can do it after committing the general introduction to "security" which I am working on

**Comment by sijie** [comment]
@eolivelli I will try to go through this again. 

/cc @merlimat  for reviewing as well

**Comment by sijie** [comment]
fyi, I added @merlimat and me to the reviewers list, making sure we have other eyes on reviewing this change.

**Comment by sijie** [comment]
@kishorekasi do you have time to review the comments? It would be great if we can attempt to close this one

**Comment by sijie** [comment]
@kishorekasi @eolivelli what is the plan for this? Are we including this in 4.5.0?

**Comment by eolivelli** [comment]
From my point of view it can be deferred to 4.6. I needed zk acls and sasl and all of these features are in

File: bookkeeper-server/bin/bookkeeper [hunk 21]
```diff
@@ -207,9 +207,9 @@ if [ $COMMAND == "bookie" ]; then
 elif [ $COMMAND == "autorecovery" ]; then
   exec $JAVA $OPTS $JMX_ARGS org.apache.bookkeeper.replication.AutoRecoveryMain --conf $BOOKIE_CONF $@
 elif [ $COMMAND == "localbookie" ]; then
-  NUMBER=$1
-  shift
-  exec $JAVA $OPTS $JMX_ARGS org.apache.bookkeeper.util.LocalBookKeeper $NUMBER $BOOKIE_CONF $@
+    NUMBER=$1
```
Comment (kishorekasi) [inline bookkeeper-server/bin/bookkeeper]: It was unintentional. I will put it back.

File: bookkeeper-server/certs/keyStoreClientPassword.txt [hunk 22]
```diff
@@ -0,0 +1 @@
+client
```
Comment (kishorekasi) [inline bookkeeper-server/certs/keyStoreClientPassword.txt]: These are sample self signed certs for SSL related tests.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingAddOp.java [hunk 23]
```diff
@@ -255,7 +255,7 @@ void submitCallback(final int rc) {
         ReferenceCountUtil.release(toSend);
 
         if (LOG.isDebugEnabled()) {
-            LOG.debug("Submit callback (lid:{}, eid: {}). rc:{}", new Object[] { lh.getId(), entryId, rc });
+            LOG.debug("Submit callback (lid:{}, eid: {}). rc:{}", lh.getId(), entryId, rc);
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/client/PendingAddOp.java]: The slf4j version we use doesn't need it anymore.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/AuthHandler.java [hunk 24]
```diff
@@ -142,11 +145,9 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
         }
 
         private boolean checkAuthPlugin(AuthMessage am, final Channel src) {
-            if (!am.hasAuthPluginName()
-                || !am.getAuthPluginName().equals(authProviderFactory.getPluginName())) {
-                LOG.error("Received message from incompatible auth plugin. Local = {},"
-                          + " Remote = {}, Channel = {}",
-                          authProviderFactory.getPluginName(), am.getAuthPluginName());
+            if (!am.hasAuthPluginName() || !am.getAuthPluginName().equals(authProviderFactory.getPluginName())) {
+                LOG.error("Received message from incompatible auth plugin. Local = {}," + " Remote = {}, Channel = {}",
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/AuthHandler.java]: oh! my editor indentation issue. I will fix that.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java [hunk 25]
```diff
@@ -509,9 +509,12 @@ public static BookieSocketAddress getBookieAddress(ServerConfiguration conf)
             iface = "default";
         }
         InetSocketAddress inetAddr = new InetSocketAddress(DNS.getDefaultHost(iface), conf.getBookiePort());
-        String hostAddress = inetAddr.getAddress().getHostAddress();
-        if (conf.getUseHostNameAsBookieID()) {
-            hostAddress = inetAddr.getAddress().getCanonicalHostName();
+        String hostAddress = "127.0.0.1";
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java]: Makes sense, '127.0.0.1' will result in a wrong address to use if unresolved.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/AuthHandler.java [hunk 26]
```diff
@@ -167,14 +168,9 @@ public void operationComplete(int rc, AuthToken newam) {
...
                 }
-                AuthMessage message =
-                    AuthMessage
-                        .newBuilder()
-                        .setAuthPluginName(req.authMessage.getAuthPluginName())
-                        .setPayload(ByteString.copyFrom(newam.getData()))
-                        .build();
-                channel.writeAndFlush(new BookieProtocol.AuthResponse(req.getProtocolVersion(),
-                                                              message));
+                AuthMessage message = AuthMessage.newBuilder().setAuthPluginName(req.authMessage.getAuthPluginName())
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/AuthHandler.java]: Again, editor indentation issue. Can be ignored.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/AuthHandler.java [hunk 27]
```diff
@@ -303,12 +304,15 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
             synchronized (this) {
                 if (authenticated) {
                     super.write(ctx, msg, promise);
+                    super.flush(ctx);
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/AuthHandler.java]: No writeAndFlush() api in channelDuplexHandler

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/AuthHandler.java [hunk 28]
```diff
@@ -303,12 +304,15 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
...
                 } else if (msg instanceof BookkeeperProtocol.Request) {
                     // let auth messages through, queue the rest
                     BookkeeperProtocol.Request req = (BookkeeperProtocol.Request) msg;
                     if (req.getHeader().getOperation()
-                            == BookkeeperProtocol.OperationType.AUTH) {
+                            == BookkeeperProtocol.OperationType.AUTH
+                        || req.getHeader().getOperation() == BookkeeperProtocol.OperationType.START_TLS) {
                         super.write(ctx, msg, promise);
+                        super.flush(ctx);
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/AuthHandler.java]: No writeAndFlush() api in channelDuplexHandler

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/AuthHandler.java [hunk 29]
```diff
@@ -317,10 +321,13 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
                     BookieProtocol.Request req = (BookieProtocol.Request)msg;
                     if (BookkeeperProtocol.OperationType.AUTH.getNumber() == req.getOpCode()) {
                         super.write(ctx, msg, promise);
+                        super.flush(ctx);
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/AuthHandler.java]: No writeAndFlush() api in channelDuplexHandler

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieNettyServer.java [hunk 30]
```diff
@@ -272,7 +363,8 @@ protected void initChannel(LocalChannel ch) throws Exception {
         }
     }
 
-    void start() {
+    void start() throws InterruptedException {
+        listenOn(bindAddress, bookieAddress);
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieNettyServer.java]: Didn't quite get it. The intention to move listenOn to start() was so that if there is an error between initialization and start() we need not bind the port and later cleanup.

File: bookkeeper-server/certs/keyStoreServerPassword.txt [hunk 31]
```diff
@@ -0,0 +1 @@
+server
```
Comment (kishorekasi) [inline bookkeeper-server/certs/keyStoreServerPassword.txt]: Unnecessary. I will remove them.

**Comment by kishorekasi** [comment]
Sije,

I have address your comments and posted a new commit. This was initially
planned for 4.5. But, I will let you prioritize it. This has gone through
multiple reviews now and it is already integrated and functional in the
Salesforce repo. It will be great if we can have it in the 4.5.

Thanks,
Kishore

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java [hunk 32]
```diff
@@ -1323,14 +1341,14 @@ private OperationType getOperationType(byte opCode) {
     private void readV3Response(final Response response) {
         final BKPacketHeader header = response.getHeader();
 
-        final CompletionValue completionValue = completionObjects.remove(newCompletionKey(header.getTxnId(),
+        final CompletionValue completionValue = completionObjects.get(newCompletionKey(header.getTxnId(),
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java]: any reason why change remove to get?

if seems a new completion key is created for each get and remove, at least if we can use same key that would reduce garbages.

**Comment by sijie** [comment]
@kishorekasi travis and jenkins failed to compile this pull request. I think a ';' is missing in ClientConfiguration. Can you take a look at it?

**Comment by kishorekasi** [comment]
@merlimat I can change SSL to TLS for class names. But rename variables to tls may not be appropriate as we will still use JDK and Netty SSL classes.

**Comment by sijie** [comment]
+1 

@kishorekasi it seems that the latest push isn't compilable. can you fix that? we are ready to merge this one.

**Comment by sijie** [comment]
merging this (will resolve the conflicts manually)
